### PR TITLE
Plugin: Crossmint Minting API

### PR DESCRIPTION
**Pull request by bryan-eastwood3**
Adds tools that allow deploying collections and minting on them with the Crossmint Minting API.
Tested and ensured that basic instructions not including details about the collection will deploy the necessary collection with generic metadata

### CODE REVIEW

File: goat.code-workspace [hunk 1]
```diff
@@ -1,67 +1,75 @@
...
-			"name": "📚 Docs",
-			"path": "./docs"
-		},
-		{
-			"name": "🚀 Examples",
-			"path": "./typescript/examples"
-		}
-	],
-	"settings": {
+    "folders": [
```
Comment (bryan-eastwood3) [inline goat.code-workspace:2]: This file was formatted incorrectly according to our biome config before

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import { z } from "zod";
+import type { Plugin, WalletClient } from "@goat-sdk/core";
+import { randomUUID } from "node:crypto";
+
+export function mintAPI(apiKey: string): Plugin<WalletClient> {
+    return {
+        name: "minting_api",
+        supportsSmartWallets: () => true,
+        supportsChain: (chain) => chain.type === "solana" || chain.type === "evm",
```
Comment (alfonso-paella) [inline typescript/packages/plugins/minting-api/src/impl.ts:9]: I think e can future roof it and add aptos here

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        getTools: async () => {
+            return [
+                {
+                    name: "create_collection",
```
Comment (alfonso-paella) [inline typescript/packages/plugins/minting-api/src/impl.ts:13]: Maybe create_nft_collection or create_token_collectio

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                    description: "This {{tool}} creates an NFT collection and returns the ID of the collection.",
+                    parameters: createCollectionParametersSchema,
+                    method: createCollectionMethod(apiKey),
+                },
+                {
+                    name: "mint_nft",
+                    description:
+                        "This {{tool}} mints an NFT to a recipient from a collection and returns the transaction hash. Requires a collection ID of an already deployed collection.",
```
Comment (alfonso-paella) [inline typescript/packages/plugins/minting-api/src/impl.ts:21]: Could we make this take a contract address instead? We support that already iirc

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import { z } from "zod";
+import type { Plugin, WalletClient } from "@goat-sdk/core";
+import { randomUUID } from "node:crypto";
+
+export function mintAPI(apiKey: string): Plugin<WalletClient> {
+    return {
+        name: "minting_api",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/minting-api/src/impl.ts:7]: I would move this whole plugin to `wallets/crossmint`. Checkout the faucet implementation: https://github.com/goat-sdk/goat/blob/main/typescript/packages/wallets/crossmint/src/faucet.ts

Then the way you get this plugin is by doing:

```typescript
const { mint } = crossmint(apiKey)
```

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            description: z.string().describe("The description of the NFT"),
+            image: z.string().describe("The image of the NFT"),
+        })
+        .describe("The metadata of the NFT"),
+});
+
+function createCollectionMethod(apiKey: string) {
+    return async (_walletClient: WalletClient, parameters: z.infer<typeof createCollectionParametersSchema>) => {
+        const id = randomUUID().toString();
+        await fetch(`https://staging.crossmint.com/api/2022-06-09/collections/${id}`, {
```
Comment (0xaguspunk) [inline typescript/packages/plugins/minting-api/src/impl.ts:63]: are we parametrizing prod and so?

File: typescript/packages/plugins/minting-api/src/impl.ts (hunk 4 above)
Comment (bryan-eastwood3) [reply typescript/packages/plugins/minting-api/src/impl.ts:21]: I don't think we do. I think we only added support for referencing NFTs by their locator

File: typescript/packages/plugins/minting-api/src/impl.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+export function mintAPI(apiKey: string): Plugin<WalletClient> {
+    return {
+        name: "minting_api",
+        supportsSmartWallets: () => true,
+        supportsChain: (chain) => chain.type === "solana" || chain.type === "evm",
+        getTools: async () => {
+            return [
+                {
+                    name: "create_collection",
+                    description: "This {{tool}} creates an NFT collection and returns the ID of the collection.",
```
[the remark this answers is not in the document]
Comment (bryan-eastwood3) [reply typescript/packages/plugins/minting-api/src/impl.ts:14]: Right now SFTs on Solana are gated to specific projects, so we can't easily support it

File: typescript/packages/plugins/minting-api/src/impl.ts (hunk 5 above)
Comment (bryan-eastwood3) [reply typescript/packages/plugins/minting-api/src/impl.ts:7]: Nice!

File: typescript/packages/wallets/crossmint/src/mintingAPI.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
+import { z } from "zod";
+import type { Plugin, WalletClient } from "@goat-sdk/core";
+import { randomUUID } from "node:crypto";
+
+export const mintingAPIFactory = (
+    apiKey: string,
+): ((options: { env: "production" | "staging" }) => Plugin<WalletClient>) => {
```
Comment (mPaella) [inline typescript/packages/wallets/crossmint/src/mintingAPI.ts:7]: i think u should just install `@crossmint/common-sdk-base`, and use `validateAPIKey` which will return the environment for the api key

then use `environmentToCrossmintBaseURL`

Comment (mPaella) [reply typescript/packages/wallets/crossmint/src/mintingAPI.ts:7]: so its not necessary to pass `env`, and its impossible to have this out of sync with the env the api key is actually for

Comment (bryan-eastwood3) [reply typescript/packages/wallets/crossmint/src/mintingAPI.ts:7]: I agree there are a lot of better ways to do this but I'm just following the pattern that already exists for plugins for the purposes of this PR

Comment (mPaella) [reply typescript/packages/wallets/crossmint/src/mintingAPI.ts:7]: theres a single other place that does it like that, and its not a breaking change to do it here, so i dont see any reason why we would not do it the correct way

Comment (0xaguspunk) [reply typescript/packages/wallets/crossmint/src/mintingAPI.ts]: Added this

**Pull request merged**